pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/Node.java

```diff
@@ -315,10 +399,48 @@
      *
      * @param childIndex
      *          The index of the child to be removed
+     *
+     * @deprecated This is internal API and will be removed from this interface with 7.0.0
      */
+    @Deprecated
+    @InternalApi
     void removeChildAtIndex(int childIndex);
 
 
+    /**
+     * Returns the parent of this node, or null if this is the {@linkplain RootNode root}
+     * of the tree.
+     *
+     * <p>This method should be preferred to {@link #jjtGetParent()}.
+     *
+     * @return The parent of this node
+     */
+    Node getParent();
+
+    /**
+     * Returns the child of this node at the given index.
+     *
+     * @throws IndexOutOfBoundsException if the index is negative or greater than {@link #getNumChildren()}.
+     */
+    Node getChild(int index);
+
+
+    /**
+     * Returns the number of children of this node.
+     */
+    int getNumChildren();
+
+    /**
+     * Returns the index of this node in its parent's children. If this
+     * node is a {@linkplain RootNode root node}, returns -1.
+     *
+     * <p>This method replaces {@link #jjtGetChildIndex()}, whose name was
+     * JJTree-specific.
+     *
+     * @return The index of this node in its parent's children
+     */
+    int getIndexInParent();
```

Comment:
I think, I once added this method. But it's hard to understand. The use case was mainly to figure out the next sibling.
Maybe we should add methods like `Node getNextSibling()` and `Node getPreviousSibling()`?
There is one more use case: Sometimes, you want to know, whether your node is the first child, which would be the case if `getIndexInParent() == 0` or `getPreviousSibling() == null`.
If you still need the index, then you could call `getParent().getChildren().indexOf(node)`. Internally, we might need to keep track of this index (to efficiently implement getNextSibling()), but I don't think, we need to explicitly publish it as an API.

Reply:
Similarly, with node streams you can do eg `node.asStream().followingSiblings()`, which uses the parent and `getIndexInParent()` to reduce this to iterating over a slice of the array, so I think it should be efficient.